Sources/Apollo/URLSessionClient.swift

```diff
@@ -247,21 +253,47 @@ open class URLSessionClient: NSObject, URLSessionDelegate, URLSessionTaskDelegat
   
   // MARK: - URLSessionDataDelegate
   
-  open func urlSession(_ session: URLSession,
-                       dataTask: URLSessionDataTask,
-                       didReceive data: Data) {
+  open func urlSession(
+    _ session: URLSession,
+    dataTask: URLSessionDataTask,
+    didReceive data: Data
+  ) {
     guard dataTask.state != .canceling else {
       // Task is in the process of cancelling, don't bother handling its data.
       return
     }
-    
-    self.$tasks.mutate {
-      guard let taskData = $0[dataTask.taskIdentifier] else {
-        assertionFailure("No data found for task \(dataTask.taskIdentifier), cannot append received data")
+
+    guard let taskData = self.tasks[dataTask.taskIdentifier] else {
```

Comment:
By removing the `self.$tasks.mutate` block, this is no longer atomic. I get why it's not necessary for updating the `TaskData` in place, because `TaskData` is a `class`. But does this mean that this is not guaranteed to happen in a sequentially? Can these create problematic race conditions?

Reply:
`tasks` is annotated with the `@Atomic` property wrapper so [all access to the task list should still be atomic](https://github.com/apollographql/apollo-ios/blob/main/Sources/Apollo/Atomic.swift#L22); we're assured that the task won't be removed at least. We only need `mutate` when we want to alter the task list.

`data` within each `TaskData` is not protected by anything though so yes I suppose there is a chance of a race condition when working with incoming data for a particular task. We shouldn't be using the list atomic lock to ensure consistent data access for a single task though. I suspect it was either a mistake to have `mutate` there in the first place or it was being abused for the purpose of single task data access.

The default callback queue is `main` which, being a serial queue, will ensure there is no race condition. Anyone who cares about a performant UI though will be giving us their own queue, which could be concurrent and introduce the potential for a race condition.

We could use `@Atomic` on `TaskData.data` to prevent this but I'd want a performance benchmark test on that to measure the cost of doing so.

Reply:
Yes, I agree with your assessment here. I do think we should be making the mutation of each individual `TaskData` be atomic though. This is far too likely to cause a race condition for someone in its current state.

Reply:
👍🏻 - created https://github.com/apollographql/apollo-ios/issues/2900 to handle it.